/small/bt_navigator:
  ros__parameters:
    use_sim_time: false
    global_frame: map
    robot_base_frame: base_link
    odom_topic: /small/odom
    bt_loop_duration: 10
    default_server_timeout: 20
    plugin_lib_names:
    - nav2_compute_path_to_pose_action_bt_node
    - nav2_follow_path_action_bt_node
    - nav2_spin_action_bt_node
    - nav2_wait_action_bt_node
    - nav2_clear_costmap_service_bt_node
    - nav2_goal_updated_condition_bt_node
    - nav2_globally_updated_goal_condition_bt_node
    - nav2_is_path_valid_condition_bt_node
    - nav2_rate_controller_bt_node
    - nav2_recovery_node_bt_node
    - nav2_pipeline_sequence_bt_node
    - nav2_round_robin_node_bt_node

/small/controller_server:
  ros__parameters:
    use_sim_time: false
    controller_frequency: 20.0
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.5
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.3
    progress_checker_plugin: "progress_checker"
    goal_checker_plugins: ["general_goal_checker"]
    controller_plugins: ["FollowPath"]
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.6
      movement_time_allowance: 10.0
    general_goal_checker:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.08
      yaw_goal_tolerance: 0.05
    FollowPath:
      plugin: "mep3_navigation::RegulatedPurePursuitController"
      desired_linear_vel: 0.5
      max_linear_accel: 0.8
      max_linear_decel: 0.8
      max_linear_jerk: 99999.0
      max_angular_jerk: 99999.0
      lookahead_dist: 0.05
      min_lookahead_dist: 0.15
      max_lookahead_dist: 0.5
      lookahead_time: 0.625
      rotate_to_heading_angular_vel: 5.0
      transform_tolerance: 0.1
      use_velocity_scaled_lookahead_dist: true
      min_approach_linear_velocity: 0.05
      use_approach_linear_velocity_scaling: true
      max_allowed_time_to_collision: 0.6
      use_regulated_linear_velocity_scaling: true
      use_cost_regulated_linear_velocity_scaling: false
      regulated_linear_scaling_min_radius: 0.2
      regulated_linear_scaling_min_speed: 0.1
      allow_reversing: true
      use_rotate_to_heading: false
      use_rotate_to_goal: true
      kp_angle: 4.0
      rotate_to_heading_min_angle: 0.785
      max_angular_accel: 5.0
      cost_scaling_dist: 0.7
      cost_scaling_gain: 1.0
      inflation_cost_scaling_factor: 5.0

/small/local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 20.0
      publish_frequency: 5.0
      global_frame: odom
      robot_base_frame: base_link
      use_sim_time: false
      rolling_window: True
      width: 3
      height: 3
      resolution: 0.05
      robot_radius: 0.20
      use_maximum: true
      plugins: ["obstacle_layer", "dilation_layer"]
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        footprint_clearing_enabled: true
        observation_sources: scan
        scan:
          topic: /small/scan
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "LaserScan"
          inf_is_valid: True
          raytrace_max_range: 3.0
          raytrace_min_range: 0.0
          obstacle_max_range: 2.5
          obstacle_min_range: 0.0
      dilation_layer:
        plugin: "mep3_navigation::DilationLayer"
        size: 0.25
        type: 2
      always_send_full_costmap: True

/small/global_costmap:
  global_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 2.0
      global_frame: map
      robot_base_frame: base_link
      use_sim_time: false
      rolling_window: True
      robot_radius: 0.2
      resolution: 0.07
      use_maximum: true
      plugins: ["obstacle_layer", "dilation_layer", "temporal_obstacle_layer", "inflation_layer"]
      dilation_layer:
        plugin: "mep3_navigation::DilationLayer"
        size: 0.25
        type: 2
      obstacle_layer:
        plugin: "nav2_costmap_2d::ObstacleLayer"
        enabled: True
        observation_sources: scan
        scan:
          topic: /small/scan
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "LaserScan"
          inf_is_valid: True
          raytrace_max_range: 3.0
          raytrace_min_range: 0.0
          obstacle_max_range: 2.5
          obstacle_min_range: 0.0
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 4.0
        inflation_radius: 1.0
      temporal_obstacle_layer:
        plugin: "mep3_navigation::TemporalObstacleLayer"
        add_obstacle_topic: /small/add_obstacle
        remove_obstacle_topic: /small/remove_obstacle
        predefined_obstacle_labels: ['side_top', 'side_bottom', 'side_left', 'side_right',
         'cherry_dispenser_top', 'cherry_dispenser_bottom', 'cherry_dispenser_left', 'cherry_dispenser_right']
        side_top: '[[1.05, -1.55], [1.05, 1.55], [1.25, 1.55], [1.25, -1.55]]'
        side_bottom: '[[-1.05, -1.55], [-1.05, 1.55], [-1.25, 1.55], [-1.25, -1.55]]'
        side_left: '[[1.05, 1.55], [-1.05, 1.55], [-1.05, 1.75], [1.05, 1.75]]'
        side_right: '[[1.05, -1.55], [-1.05, -1.55], [-1.05, -1.75], [1.05, -1.75]]'
        cherry_dispenser_top: '[[-0.05, 1.5], [0.05, 1.5], [0.05, 1.15], [-0.05, 1.15]]'
        cherry_dispenser_bottom: '[[-0.05, -1.5], [0.05, -1.5], [0.05, -1.15], [-0.05, -1.15]]'
        #cherry_dispenser_left: '[[-0.95, 0.1], [-0.85, 0.1], [-0.95, -0.1], [-0.95, 0.1]]'
        #cherry_dispenser_right: '[[0.95, 0.1], [0.85, 0.1], [0.95, -0.1], [0.95, 0.1]]'
      always_send_full_costmap: True

/small/planner_server:
  ros__parameters:
    use_sim_time: false
    planner_plugins: ["GridBased"]
    expected_planner_frequency: 1.0
    GridBased:
      plugin: 'nav2_navfn_planner/NavfnPlanner'
      use_astar: True
      allow_unknown: false
      use_final_approach_orientation: False
      tolerance: 0.025

/small/behavior_server:
  ros__parameters:
    costmap_topic: local_costmap/costmap_raw
    footprint_topic: local_costmap/published_footprint
    cycle_frequency: 100.0
    behavior_plugins: ["spin", "wait", "move"]
    spin:
      plugin: "nav2_behaviors/Spin"
    wait:
      plugin: "nav2_behaviors/Wait"
    move:
      plugin: "mep3_navigation::MoveBehavior"
    global_frame: odom
    robot_base_frame: base_link
    transform_tolerance: 0.1
    use_sim_time: false
    max_rotational_vel: 1.0
    min_rotational_vel: 0.4
    rotational_acc_lim: 3.2
    simulate_ahead_time: 0.3
    
    # Move
    simulate_ahead_distance: 0.2
    debouncing_duration: 0.05
    linear:
      max_velocity: 1.2
      max_acceleration: 1.0
      kp: 5.0
      tolerance: 0.01
    angular:
      max_velocity: 8.0
      max_acceleration: 5.0
      kp: 8.0
      tolerance: 0.01
